ophyd/areadetector/filestore_mixins.py

```diff
@@ -241,25 +247,66 @@ def write_path_template(self, val):
         self._write_path_template = _ensure_trailing_slash(val)
 
     def stage(self):
-        self._point_counter = count()
+        self._datum_counter = count()
         self._locked_key_list = False
         self._datum_uids.clear()
         super().stage()
         if self.enable.get() and self._reg is None:
             raise RuntimeError('The plugin {!r} is enabled, but '
                                'the Registry (self._reg) is `None`')
+        # Subclasses will assemble resource_kwargs and call
+        # self._generate_resource(resource_kwrags)
+
+    def _generate_resource(self, resource_kwargs):
+        fn = PurePath(self._fn).relative_to(self.reg_root)
+        resource = {'spec': self.filestore_spec,
+                    'root': str(self.reg_root),
+                    'resource_path': str(fn),
+                    'resource_kwargs':resource_kwargs,
+                    'path_semantics': self.path_semantics}
+        # If a Registry is set, we need to allow it to generate the uid for us.
+        if self._reg is not None:
+            logger.debug("Inserting resource with filename %s", self._fn)
+            # register_resource has accidentally different parameter names...
+            self._resource_uid = self._reg.register_resource(
+                rpath=resource['resource_path'],
+                rkwargs=resource['resource_kwargs'],
+                root=resource['root'],
+                spec=resource['spec'],
+                path_semantics=resource['path_semantics'])
+        # If a Registry is not set, we need to generate the uid.
+        else:
+            self._resource_uid = new_uid()
+        resource['uid'] = self._resource_uid
+        self._asset_docs_cache.append(('resource', resource))
 
     def generate_datum(self, key, timestamp, datum_kwargs):
         "Generate a uid and cache it with its key for later insertion."
+        i = next(self._datum_counter)
         datum_kwargs = datum_kwargs or {}
+        datum_kwargs.update({'point_number': i})
```

Comment:
wait, **every** file handler will get a `point_number` now? Is that safe?
For example, here in `eiger_io`:
https://github.com/NSLS-II/eiger-io/blob/master/eiger_io/fs_handler.py#L142

If I had to vote, I'd say it's worth the risk, but it's something we're going to have to bring attention to the whole group, and monitor these next few weeks (upgrade handlers to take in a `point_number`). I may have misunderstood something though. It might be good for another fresh pair of eyes to review this.
Here is Eiger's file plugin, inheriting from `FileStoreBase` [here](https://github.com/NSLS-II-CHX/profile_collection/blob/master/startup/20-area-detectors.py#L67).

Reply:
Good catch!  This should be moved back to the sub-class it was pulled up out of.  The eiger filestore plugin must inherit from the base class rather than `FileStoreIterativeWrite` like most of the other plugins.

Reply:
that could be a problem since there will be two dependencies on the iterator `datum_counter`.
Actually, I don't know if I like `{resource_id}/{point_number}` as the uid. Can we just have it be a uid?
What if a user decides they want  `_datum_counter` to be a sequence id or do something else funny?
Or what if the user finds a way to use the same resource for multiple `stage/unstage` runs?

I understand we want to keep the structure of the read datums, so how about:
`{resource_uid}/{datum_uid}` where `datum_uid=str(uuid.uuid4()[:10])` or something?

Reply:
The current 'status-quo' is just `str(uuid.uuid4())` but as part of adding the 'column_hdf5' asset registry and the performance work around that (https://github.com/NSLS-II/databroker/pull/216 for the code, did not keep good notes on the performance work) we are starting to move to `datum_id = {resourc_uid}/{N}` to a) support random and bulk access in column case and b) generating the uuids for each datum was starting to contribute significantly to run time for lots of datum (on my laptop it is 6 µs per uuid which for a 2k by 2k scan is ~24 seconds total time)